packages/assets-controllers/src/TokenDetectionController.test.ts

```diff
@@ -120,33 +144,117 @@ const setupTokenListController = (
   });
 
   return { tokenList, tokenListMessenger };
-};
+}
+
+/**
+ * Builds a messenger that `TokenDetectionController` can use to communicate with other controllers.
+ * @param controllerMessenger - The main controller messenger.
+ * @returns The restricted messenger.
+ */
+function buildTokenDetectionControllerMessenger(
+  controllerMessenger: MainControllerMessenger = getControllerMessenger(),
+): TokenDetectionControllerMessenger {
+  return controllerMessenger.getRestricted({
+    name: controllerName,
+    allowedActions: [
+      'NetworkController:getNetworkClientById',
+      'TokensController:getState',
+      'TokensController:addDetectedTokens',
+      'TokenListController:getState',
+    ],
+    allowedEvents: [
+      'NetworkController:stateChange',
+      'NetworkController:networkDidChange',
+      'TokenListController:stateChange',
+    ],
+  });
+}
 
 describe('TokenDetectionController', () => {
   let tokenDetection: TokenDetectionController;
   let preferences: PreferencesController;
   let tokensController: TokensController;
+  let tokensMessenger: TokensControllerMessenger;
   let tokenList: TokenListController;
   let controllerMessenger: MainControllerMessenger;
   let getBalancesInSingleCall: sinon.SinonStub<
     Parameters<AssetsContractController['getBalancesInSingleCall']>,
     ReturnType<AssetsContractController['getBalancesInSingleCall']>
   >;
 
-  const onNetworkDidChangeListeners: ((state: NetworkState) => void)[] = [];
+  const networkDidChangeEvent = jest.fn<NetworkState, []>();
+  let getNetworkClientByIdHandler = jest.fn();
+  const getNetworkClientByIdInfuraHandler = jest.fn<
+    AutoManagedNetworkClient<InfuraNetworkClientConfiguration>,
+    Parameters<NetworkController['getNetworkClientById']>
+  >();
+  const getNetworkClientByIdCustomHandler = jest.fn<
+    AutoManagedNetworkClient<CustomNetworkClientConfiguration>,
+    Parameters<NetworkController['getNetworkClientById']>
+  >();
   const changeNetwork = (providerConfig: ProviderConfig) => {
-    onNetworkDidChangeListeners.forEach((listener) => {
-      listener({
-        ...defaultNetworkState,
-        providerConfig,
-      });
+    networkDidChangeEvent.mockReturnValue({
+      ...defaultNetworkState,
+      providerConfig,
+      selectedNetworkClientId: providerConfig.id ?? providerConfig.type,
     });
```

Comment:
It seems you're merely modifying the return value of this handler, but you're never re-issuing the `networkDidChange` event. Does that need to happen here?

``` typescript
    controllerMessenger.publish('NetworkController:networkDidChange', {
      ...defaultNetworkState,
      providerConfig,
      selectedNetworkClientId: providerConfig.id ?? providerConfig.type,
    });
```

Reply:
Hmm I did add this publish call (71d9967d3559111d528bf279d593167ed8e82409), but then I removed it at some point. I'm not exactly sure why, but I think I misunderstood this suggestion: https://github.com/MetaMask/core/pull/3609#discussion_r1424681019. 

Restored here! 5a94e14cf51472d39be5a3f3a0fc08f989a5af5e